### ISSUE TITLE
Matter Update possible?

### ISSUE THREAD

**Issue by derpate1989** [Aircoookie/WLED#2871]
Hi, is a Matter Update possible? :) That would be a huge gain for Wled and various smart home systems.

**Comment by Aircoookie** [comment]
Matter support would definitely be great as it might just be the missing link to make WLED devices work locally with all common smart speakers.
Right now I see three main issues, which I hope can be solved:
- There should be an Arduino library for it ([this](https://github.com/jakubdybczak/esp32-arduino-matter) seems like a starting point already)
- It requires encryption, and unless the matter specification allows self-signed certs (which I doubt), it will be a pain to have a PKI infrastructure. Also, due to memory constraints, this will rule out matter ever working on ESP8266.
- I am not sure whether Bluetooth provisioning is optional or a core requirement of the Matter specification. Bluetooth on ESP32 is supported, but uses an extreme amount of memory and storage, which is why we steered clear of using it within WLED thus far.

**Comment by derpate1989** [comment]
Thank you for this detailed explanation! I have a lot of hope in Matter and look forward to this standard. Maybe someday there will be a way for Wled and Matter :)

**Comment by DoomHammer** [comment]
> Matter support would definitely be great as it might just be the missing link to make WLED devices work locally with all common smart speakers. Right now I see three main issues, which I hope can be solved:
> 
> * There should be an Arduino library for it ([this](https://github.com/jakubdybczak/esp32-arduino-matter) seems like a starting point already)
> * It requires encryption, and unless the matter specification allows self-signed certs (which I doubt), it will be a pain to have a PKI infrastructure. Also, due to memory constraints, this will rule out matter ever working on ESP8266.
> * I am not sure whether Bluetooth provisioning is optional or a core requirement of the Matter specification. Bluetooth on ESP32 is supported, but uses an extreme amount of memory and storage, which is why we steered clear of using it within WLED thus far.
> 
> Still I am enthusiastic for the Matter standard, if only for realizing that a unified smart home standard is necessary and that mandatory device vendor clouds are a terrible idea.

If it's not possible to do entirely on a device, maybe a bridge could help here?

**Comment by Aircoookie** [comment]
That's a nice idea, it should definitely be feasible to make some software to act as a bridge between the JSON API and Matter protocol(s), and it would even support ESP8266 units. Native support would be ideal though.

**Comment by jasta** [comment]
> - I am not sure whether Bluetooth provisioning is optional or a core requirement of the Matter specification. Bluetooth on ESP32 is supported, but uses an extreme amount of memory and storage, which is why we steered clear of using it within WLED thus far.
> 

I don't believe it is.  The spec says that there are 3 methods, BLE, WiFi SoftAp (which no one is using today unfortunately), but also existing DNS-SD if you're already on the local network.  This last case seems perfect for WLED which is already so easy to get setup to connect to WiFi.

It seems that it just needs to register some mDNS records like _matter._udp/_tcp maybe?  I'm still trying to get full access to the spec to read the details myself and start playing around with this a bit.

**Comment by Chase295** [comment]
Is there any new stuff or soultion a "bridge" or any think else becous i would like to add it in the nanoleaf app for the new Nanoleaf 4D

**Comment by blazoncek** [comment]
Unless someone with knowledge and programming skills joins in there will be no progress.

**Comment by heho0011** [comment]
Hi,
maybe Matter can be implemented the same way as in tasiota for esp32?

**Comment by Stoffe** [comment]
Seems like Espressif has a new esp32 with support?
https://www.espressif.com/en/news/ESP32-C6_Available

And using Yacubane esp32 lib, you could maybe implement matter in esp32 or take ideas from...
https://github.com/Yacubane/esp32-arduino-matter
Here is a video of the install, https://www.youtube.com/watch?v=HpACspFygUI

might be something to look into...
I would also upvote this feature for sure.

**Comment by marcoscv-work** [comment]
I have seen that tasmota already has support for Matter. I've been testing and works fine in homekit. it's interesting because they provide ESP8266 support by ESP32, Looks promising and a really good idea.
https://tasmota.github.io/docs/Matter/

Right now all solutions to integrate WLED in homekit are pretty painful, implementing matter in some way would be awesome.

**Comment by jonahnm** [comment]
I will be attempting to implement this soon.

**Comment by khimaros** [comment]
@Aircoookie thank you for the detailed breakdown, a few questions inline:

>     * There should be an Arduino library for it ([this](https://github.com/jakubdybczak/esp32-arduino-matter) seems like a starting point already)

what sort of signal would you look for in this library to consider it ready for use in WLED?

i filed https://github.com/Yacubane/esp32-arduino-matter/issues/44 re: publishing to platformio registry, but i'm not sure if you consider this a blocker.

>     * It requires encryption, and unless the matter specification allows self-signed certs (which I doubt), it will be a pain to have a PKI infrastructure. Also, due to memory constraints, this will rule out matter ever working on ESP8266.

from what i've been able to find, there is no restriction placed on which CAs are used or any other verbiage limiting use of self-signed certificates in the spec.

if this were restricted, a workaround is to use LetsEncrypt with real domain name the user controls, or to trust/use a services such as https://sslip.io/ or https://backname.io/ to get valid certs with any IP address (including private blocks).

>     * I am not sure whether Bluetooth provisioning is optional or a core requirement of the Matter specification. Bluetooth on ESP32 is supported, but uses an extreme amount of memory and storage, which is why we steered clear of using it within WLED thus far.

the library you linked (https://github.com/jakubdybczak/esp32-arduino-matter) "comes with precompiled NimBLE, because default Bluedroid shipped with arduino-esp32 takes too much RAM memory", does that resolve your concern?

> Still I am enthusiastic for the Matter standard, if only for realizing that a unified smart home standard is necessary and that mandatory device vendor clouds are a terrible idea.

it would be awesome for users to be able to use their existing apps and assistants to also control their WLED devices. great for spouse approval of these devices :)

**Comment by chikko80** [comment]
@khimaros @Aircoookie 

[https://github.com/espressif/arduino-esp32/pull/10467](https://github.com/espressif/arduino-esp32/pull/10467)

Maybe i am wrong here, i am relatively new to this space, but i stumbled across this merge request from [espressif](https://github.com/espressif) which was merged very recently and seems to add matter support for esp32.
Can you work with that?

**Comment by softhack007** [comment]
@chikko80 thanks for the hint 👍 

Its PR for arduino-esp32 v3.1, so unfortunately it's still far away from us - we are using v2.0.9 right now for some boards.  

At the moment we cannot upgrade a newer arduino version for several reasons
* bugs in the new arduino framework - espressif needed 2 years to make v2.0.x usable and stable, then declared it unsupported once it was usable. So I expect the same for v3.1 which is not released yet.
* WLED does not compile with the new arduino framework v3.0.x - to many breaking changes introduced by espressif
* [pPlatformIO does not support the new v3.0.x arduino framework](https://github.com/platformio/platform-espressif32/issues/1225), and it's questionable if this will change.


But time will tell, and maybe in a few years from now we can use matter support provided by arduino-esp32 v3.1.x.

**Comment by jonahnm** [comment]
> I will be attempting to implement this soon.

And then I forgot!

**Comment by khimaros** [comment]
there is a community supported PlatformIO module with support for arduino core 3.0.x: https://github.com/pioarduino/platform-espressif32 -- see also https://github.com/espressif/arduino-esp32/discussions/10039

**Comment by quickquest88** [comment]
> > I will be attempting to implement this soon.
> 
> And then I forgot!

Did you remember this time?  In either case, look, i'm still very much a programming newb, with very little knowledge, but I can learn quickly if given direction, and I am motivated to help, I really would like to see WLED over thread. I will offer any help I can.  I;m going to try and start reading up on the thread spec and libs, and get familiar, i hope.  please let me know if I can help.

**Comment by M4rkoV2** [comment]
Any new progress on this?

**Comment by netmindz** [comment]
Not directly on Matter, but there is progress on moving to newer versions of the platform which helps move us towards being able to support things like the C6 and Matter

**Comment by mateuszstalmach** [comment]
Hi,

I saw that the latest Matter specification (1.4.2) no longer requires Bluetooth commissioning.

Would it be possible to add Matter support to WLED for ESP32 now, using only Wi-Fi commissioning?

I think this update could make Matter integration easier without extra hardware.

**Comment by netmindz** [comment]
The Espressif Matter SDK needs the V5 but we are only on V4 at the moment, so this would be blocked by 
https://github.com/wled/WLED/pull/4838

**Comment by netmindz** [comment]
See comment on how to add https://github.com/wled/WLED/issues/3078#issuecomment-3506845307

**Comment by elsiehupp** [comment]
I just want to chime in and ask: what about Matter-over-Thread?

Many ESP32 Zigbee devices [alternately support Thread](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_openthread.html), so if there is hardware support for Thread it would be nice if WLED could support Matter-over-Thread, not just Matter-over-WiFi.

**Comment by netmindz** [comment]
@elsiehupp I've not yet looked, but not sure what the advantages of Matter over Thread would be for WLED over ZigBee

I have a PoC for ZigBee support using C6 https://github.com/wled/WLED/pull/5457

Another PoC  that does Matter over WiFi and uses esp32 https://github.com/wled/WLED/pull/5456

I also have ZigBee and Matter over WiFi bridges to enable all your existing WLED installs for those ecosystems

https://github.com/netmindz/zigbee_wled

https://github.com/netmindz/matter_wifi_wled

**Comment by elsiehupp** [comment]
> I've not yet looked, but not sure what the advantages of Matter over Thread would be for WLED over ZigBee

Matter-over-Thread and Zigbee have similar features but are not directly compatible. For instance, [Matter-over-Thread supports binding](https://developer.espressif.com/blog/matter-device-to-device-automations/) in much the same way as [Zigbee](https://docs.silabs.com/zigbee/9.0.0/zigbee-concepts-network/binding), though Matter-over-Thread binding is not particularly well supported as of yet.

As for Matter-over-Thread versus Matter-over-WiFi, Thread is a mesh network designed for larger numbers of devices than most consumer WiFi routers, so in addition to supporting peer-to-peer binding Matter-over-Thread should also be more reliable at scale.

And as for bridging Matter to WLED... Matter-over-Thread with its peer-to-peer binding and mesh topology is just a lot more reliable. If individual devices can support Matter-over-Thread and WLED's own protocol simultaneously, though, all the better!

Note that I don't personally have any WLED devices, as I have been working with Zigbee (via Zigbee2MQTT), despite its disadvantages, largely because of Zigbee's similar benefits versus WiFi when compared to Matter-over-Thread. I found my way here to this Issue because my brother and I are in the process of setting up Matter-over-Thread lighting at our parents' house, and the options for Matter-over-Thread LED controllers, i.e. not just bulbs, are still somewhat limited compared to the options for Zigbee LED controllers. If some WLED LED controllers could directly support Matter-over-Thread (or even Zigbee), that support would open up a lot more options for our new setup.

**Comment by elsiehupp** [comment]
> I have a PoC for ZigBee support using C6 [#5457](https://github.com/wled/WLED/pull/5457)

I find it interesting that you're apparently taking the approach of bridging WLED to Zigbee using a bridge rather than running a Zigbee node on each and every WLED device. This approach loses the advantages of Zigbee's mesh topology and peer-to-peer bindings.

I didn't know it was even possible for a single device to present itself over Zigbee as multiple separate devices, considering how dumb and inflexible Zigbee tends to be (albeit dumb and inflexible mostly in a good way, as it cuts down on OEM shenanigans).

If individual WLED devices could simultaneously also support Zigbee, that would actually be really cool, as well, considering WLED's and Zigbee's different feature sets and advantages.

(Note: It used to be called ZigBee, but at some point it got renamed to Zigbee. Yes, it's silly they changed it. Yes, it still confuses autocorrect on my phone. Though to be fair using the form "Zigbee" mostly just communicates that the speaker is drinking the Kool-Aid enough to have noticed the change in the first place, either that or the speaker is new enough to Zigbee to have missed out on ZigBee back in the day.)

**Comment by netmindz** [comment]
@elsiehupp you have have not looked properly at what I shared. The specific pr you quoted is exactly a ZigBee node running on a WLED device, so it's just seen like any other RGB bulb

Yes it's perfectly possible to present multiple endpoints, that's exactly how my bridge works. 

You also don't lose and if the advantages of the mesh networking as in either setup you are running as a ZigBee mesh node, the only difference is what else you are running on the same hardware. 

Sharing a single radio between WiFi and ZigBee restricts WiFi performance, so has a negative impact on WLED usability, so it's better to have a fully functional WLED install but the convenience of being able to turn all you lights in the room on/off etc with zigbee

**Comment by elsiehupp** [comment]
> You also don't lose any of the advantages of the mesh networking as in either setup you are running as a ZigBee mesh node, the only difference is what else you are running on the same hardware. 

Yes and no. If you have a single WLED node acting as the Zigbee bridge and talking to other WLED nodes over WiFi, reliability would still be constrained by the reliability and scalability of the WiFi network itself.

Sure, for most purposes, WiFi is reliable enough, but WiFi does not have a mesh topology, and consumer WiFi routers are less scalable than Zigbee.

> Sharing a single radio between WiFi and ZigBee restricts WiFi performance, so has a negative impact on WLED usability, so it's better to have a fully functional WLED install but the convenience of being able to turn all you lights in the room on/off etc with zigbee

I mean… yes. Though wouldn’t Thread be a way to have it both ways?

Unlike Zigbee, Thread isn’t limited to a cluster library that hasn’t been updated since 2019. Not that this has stopped Aqara from, as I understand it, creating proprietary ZCL extensions.

So either (a) WLED could create its own proprietary (open but not standards-defined) ZCL extensions (like Aqara) and/or (b) WLED could optionally use UDP-over-Thread alongside Matter, also using UDP-over-Thread.

Creating proprietary ZCL extensions seems like a hassle, and waaaaaay beyond the scope of what I imagine WLED wanting to deal with, but running both WLED UDP-over-Thread and Matter UDP-over-Thread would allow for the reliability and scalability of a mesh network without the interference issues that come from trying to use WiFi over the same radio.

(There are devices out there that simultaneously support both Zigbee and Matter-over-Thread, but I don’t know what sort of software or hardware mitigations they use to avoid interference.)

Hence: Thread!
